### ISSUE TITLE
Improve diagram layouting

### ISSUE THREAD

**Issue by amolenaar** [gaphor/gaphor#3549]
### Is your feature request related to a problem?

Currently we have an half-baked auto layout function based on GraphViz. Most diagram types produce okay results, but diagrams like sequence diagrams are not supported.

I picked Graphviz, since it's the most common one and the results are predictable.

### Describe the solution you'd like

I came across http://blockdiag.com by accident. It looks like a good fit for our problem. Even more since everything is in Python, so it should be a lot easier to integrate: no intermediate text files.

Downside is that the code has not seen any love in around 3 years.

### Describe alternatives you've considered

Stick with Graphviz?

---
This issue is a follow-up of #1653.

An incomplete of diagram layout libraries for Python is available at https://github.com/gaphor/gaphor/issues/615#issuecomment-1054236210.

**Comment by tompkins-ct** [comment]
Improved diagram autolayout would be nice.

I think we'd want a couple capabilities to facilitate easier use
1. separate engines based on diagram type  (selectable or auto)
    1. Sequence / swimlanes probably need a different layout engine / settings.
    2. block diagrams showing functional flow vs block diagrams showing taxonomy (tree / hierarchical)
1. clean up connections only vs blocks as a whole
2. only layout selected items (limit scope of the layout engine)


I think we have a couple options here. 
1. Maybe take a look at blockdiag and re-implement its layout engine in gaphor (it's layout engine might be graphviz, but I didn't look).
    1. We may be able to improve graphviz for specific diagrams by adjusting some of the layout parameters.
3. Use the eclipse layout kernal (ELK). (https://github.com/kieler/elkjs / ). This may require some import / export work to interface with it but it's got layout engines suitable for UML/sysML type diagrams. It looks like it can import a graph via JSON or custom formats and support laying out ports and hierarchical nested systems. 
4. Implement a pure-python layout engine in gaphor (or a new library). Right now, it's doesn't look like there are a lot out there outside of graphviz. However, the algorithms on a lot of the java / js libraries are easy to access we can also take a look at journal articles for references.

**Comment by amolenaar** [comment]
We can now easily separate the layout engine based on diagram type (#3649).

I do not feel like reaching out to Java/nodejs applications. They're very hard to package in the installers we provide for Windows and macOS.

Re-implementing is a huge effort, though.

**Comment by tompkins-ct** [comment]
I fully agree that re-implementing is not a small task. From what I can tell, a lot of these layout algorithms can be more computationally intensive, so raw python may or may not be a great choice for implementing them unless you can efficiently offload the computation to something else. 

I've been playing around with using Elkjs on Gaphor and its output seems to layout nicer than dot/graphviz (especially w/ ports), but I have no feel for how hard it is to package elkjs (I also don't know how its license interacts w/ Gaphors).


Heirarchical graphs:
Graphviz: ![image](https://github.com/user-attachments/assets/e1a474fa-f530-4d84-9bda-31160d04053e)

Elkjs: 
![image](https://github.com/user-attachments/assets/d0ce930d-2d83-44fe-bb88-fd90543d6ea8)

ports:
Dot/GraphViz: 
![image](https://github.com/user-attachments/assets/3e058b14-5a73-4b7b-840a-ccdb50921a0b)

Elkjs: (I didn't include port labels in the autolayout plugin so they are overlapping...but that is an easy feature to add...)
Note: Elk references the connector endpoint as the center of the port. Gaphor seems to pick the attatchement point to be the closest to the end? When the point is in the center it leads to some odd connections. It may be useful to update how those presentations are handled in gaphor to change to join at the incoming edge?
![image](https://github.com/user-attachments/assets/6699b577-ee9d-4edd-9c4e-d39e1ffb8a99)

**Comment by amolenaar** [comment]
I quickly tried blockdiag. The results are not ideal:

![image](https://github.com/user-attachments/assets/5aede56c-4599-44d4-b436-7bcff936e35c)


<details>
<summary>blockdiag markup</summary>
```
{
 test1 -> 2.2;
 test1 -> test2;
 test3 -> test2;
 2.3 -> test2;
 2.4 -> aklsjd;
 BB -> 2.2;
  
 group hierarchical_node {
  shape = line;
  color = blue;
  
  group container_2 {
   shape = line;
   color = red;
   2.1 -> 2.3;
   2.2 -> 2.4;
   2.1, 2.2, 2.3, 2.4;
  }
  group container_3 {
   shape = line;
   color = green;
   test1, test2, test3;
  }
 }
 group big_container {
   shape = line;
  "New Block", aklsjd, BB
 }
}
```
</details>

**Comment by tompkins-ct** [comment]
I can make a PR so you can take a look. 

1. I used the current layout plugin as a basis for extracting a graph from the diagram of interest. 
2. Converted to JSON and passed to a small node script via `subprocess` 
3. the node script just loads ELKjs, passes the JSON string, and returns the formatted graph
4. python reads the output and updates the diagram

One issue is that ELK doesn't handle edge-to-edge connections, so we may need to play some tricks to get that to work.
- option 1: maybe add a 0 sized node to the output graph to represent the connection. We'd have to split an edge and then recombine it on the python size
- option 2: combine the 2 Gaphor edges into a hyperedge and then split the resulting hyperedge from elk. (not all of the elk layout algorithms support hyperedges)

One thing that would be helpful from gaphor would be a diagram showing how things are laid out. Elk provides [this](https://eclipse.dev/elk/documentation/tooldevelopers/graphdatastructure/coordinatesystem.html), which was helpful to make sure everything was correctly placed. I just assumed gaphor was referenced to the top left corner for all items (which is mostly true, but ports seem to be an exception because they are located based on the center coordinate.)
